### ISSUE TITLE
qubes-network-uplink@eth0.service fails at VM startup if vCPU is 1

### ISSUE THREAD

**Issue by noskb** [QubesOS/qubes-issues#8362]
[How to file a helpful issue](https://www.qubes-os.org/doc/issue-tracking/)

### Qubes OS release

R4.2

### Brief summary
As title. A workaround is to increase vcpu to 2 or more, downgrade core-agent-linux to v4.2.12, or revert https://github.com/QubesOS/qubes-core-agent-linux/commit/736803f400e0d3d728aba2999bc73b00bd20cdc7.

### Steps to reproduce

Start AppVM with vcpu set to 1

### Expected behavior

qubes-network-uplink<span>@</span>eth0.service is successfully and has network access.

### Actual behavior

qubes-network-uplink<span>@</span>eth0.service failed and has no network access.

<details>
<summary>sudo systemctl status qubes-network-uplink@eth0.service</summary>

```
× qubes-network-uplink@eth0.service - Qubes network uplink (eth0) setup
     Loaded: loaded (/usr/lib/systemd/system/qubes-network-uplink@.service; static)
     Active: failed (Result: exit-code) since Sat 2023-07-15 09:44:18 GMT; 1min 36s ago
   Duration: 670ms
    Process: 567 ExecStart=/usr/lib/qubes/setup-ip add eth0 (code=exited, status=2)
   Main PID: 567 (code=exited, status=2)
        CPU: 17ms

Jul 15 09:44:18 disp884 systemd[1]: Starting qubes-network-uplink@eth0.service - Qubes network uplink (eth0) setup...
Jul 15 09:44:18 disp884 setup-ip[582]: RTNETLINK answers: File exists
Jul 15 09:44:18 disp884 systemd[1]: qubes-network-uplink@eth0.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Jul 15 09:44:18 disp884 systemd[1]: qubes-network-uplink@eth0.service: Failed with result 'exit-code'.
Jul 15 09:44:18 disp884 systemd[1]: Failed to start qubes-network-uplink@eth0.service - Qubes network uplink (eth0) setup.
```
</details>

**Comment by noskb** [comment]
I can reproduce it on fedora-37 and debian-11, but not on fedora-38 and debian-12.

**Comment by noskb** [comment]
I disabled [qubes-network-uplink.service](https://github.com/QubesOS/qubes-core-agent-linux/blob/v4.2.14/vm-systemd/qubes-network-uplink.service) that calls [network-uplink-wait.sh](https://github.com/QubesOS/qubes-core-agent-linux/blob/v4.2.14/vm-systemd/network-uplink-wait.sh) and the problem no longer occurs. Is it possibly relevant?

**Comment by kennethrrosen** [comment]
I'm also having this, though in a F38 template, in the appVM I use as a VPN proxy. Disabling `[qubes-network-uplink.service](https://github.com/QubesOS/qubes-core-agent-linux/blob/v4.2.14/vm-systemd/qubes-network-uplink.service)' does not work.

```
× qubes-network-uplink@eth0.service - Qubes network uplink (eth0) setup
     Loaded: loaded (/usr/lib/systemd/system/qubes-network-uplink@.service; static)
    Drop-In: /usr/lib/systemd/system/service.d
             └─10-timeout-abort.conf
     Active: failed (Result: exit-code) since Tue 2023-10-10 08:12:49 EST; 55s ago
    Process: 500 ExecStart=/usr/lib/qubes/setup-ip add eth0 (code=exited, status=1/FAILURE)
   Main PID: 500 (code=exited, status=1/FAILURE)
        CPU: 170ms
```
and
```
× qubes-network-uplink.service - Qubes network uplink wait
     Loaded: loaded (/usr/lib/systemd/system/qubes-network-uplink.service; enabled; preset: enabled)
    Drop-In: /usr/lib/systemd/system/service.d
             └─10-timeout-abort.conf
     Active: failed (Result: exit-code) since Tue 2023-10-10 08:12:49 EST; 1min 23s ago
    Process: 425 ExecStart=/usr/lib/qubes/init/network-uplink-wait.sh (code=exited, status=1/FAILURE)
   Main PID: 425 (code=exited, status=1/FAILURE)
        CPU: 60ms

```
The error I get when attempting to `enable qubes-network-uplink` or 'disable qubes-network-uplink` is:
```
The unit files have no installation config (WantedBy=, RequiredBy=, Also=,
Alias= settings in the [Install] section, and DefaultInstance= for template
units). This means they are not meant to be enabled or disabled using systemctl.
 
Possible reasons for having this kind of units are:
• A unit may be statically enabled by being symlinked from another unit's
  .wants/ or .requires/ directory.
• A unit's purpose may be to act as a helper for some other unit which has
  a requirement dependency on it.
• A unit may be started when needed via activation (socket, path, timer,
  D-Bus, udev, scripted systemctl call, ...).
• In case of template units, the unit is meant to be enabled with some
  instance name specified.
```
And here's the output from `journalctl -xeu qubes-network-uplink.service`:
```
Oct 10 08:12:49 core-vpn systemd[1]: qubes-network-uplink@eth0.service: Main process exited, code=exited, status=1/FAILURE
░░ Subject: Unit process exited
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░ 
░░ An ExecStart= process belonging to unit qubes-network-uplink@eth0.service has exited.
░░ 
░░ The process' exit code is 'exited' and its exit status is 1.
Oct 10 08:12:49 core-vpn systemd[1]: qubes-network-uplink@eth0.service: Failed with result 'exit-code'.
░░ Subject: Unit failed
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░ 
░░ The unit qubes-network-uplink@eth0.service has entered the 'failed' state with result 'exit-code'.
Oct 10 08:12:49 core-vpn systemd[1]: Failed to start qubes-network-uplink@eth0.service - Qubes network uplink (eth0) setup.
░░ Subject: A start job for unit qubes-network-uplink@eth0.service has failed
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░ 
░░ A start job for unit qubes-network-uplink@eth0.service has finished with a failure.
░░ 
░░ The job identifier is 141 and the job result is failed.
```

**Comment by kennethrrosen** [comment]
```
Oct 10 08:32:39 core-vpn systemd[1]: qubes-network-uplink@eth0.service: Main process exited, code=exited, status=1/FAILURE
░░ Subject: Unit process exited
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░ 
░░ An ExecStart= process belonging to unit qubes-network-uplink@eth0.service has exited.
░░ 
░░ The process' exit code is 'exited' and its exit status is 1.
Oct 10 08:32:39 core-vpn systemd[1]: qubes-network-uplink@eth0.service: Failed with result 'exit-code'.
░░ Subject: Unit failed
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░ 
░░ The unit qubes-network-uplink@eth0.service has entered the 'failed' state with result 'exit-code'.
Oct 10 08:32:39 core-vpn systemd[1]: Failed to start qubes-network-uplink@eth0.service - Qubes network uplink (eth0) setup.
░░ Subject: A start job for unit qubes-network-uplink@eth0.service has failed
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░ 
░░ A start job for unit qubes-network-uplink@eth0.service has finished with a failure.
░░ 
░░ The job identifier is 141 and the job result is failed.

```
`qubes-core-agent` is: qubes-core-agent-4.2.21-1.fc38.x86_64

I can ping 1.1.1.1 successfully and run dnf update, but attempting to downgrade the package yields:
```
Total download size: 270 k
Is this ok [y/N]: y
Downloading Packages:
[MIRROR] qubes-core-agent-nautilus-4.2.20-1.fc38.x86_64.rpm: Curl error (7): Couldn't connect to server for https://yum.qubes-os.org/r4.2/current/vm/fc38/rpm/qubes-core-agent-nautilus-4.2.20-1.fc38.x86_64.rpm [Failed to connect to yum.qubes-os.org port 443 after 275 ms: Couldn't connect to server]
[MIRROR] qubes-core-agent-4.2.20-1.fc38.x86_64.rpm: Curl error (7): Couldn't connect to server for https://yum.qubes-os.org/r4.2/current/vm/fc38/rpm/qubes-core-agent-4.2.20-1.fc38.x86_64.rpm [Failed to connect to yum.qubes-os.org port 443 after 279 ms: Couldn't connect to server]
[MIRROR] qubes-core-agent-dom0-updates-4.2.20-1.fc38.noarch.rpm: Curl error (7): Couldn't connect to server for https://yum.qubes-os.org/r4.2/current/vm/fc38/rpm/qubes-core-agent-dom0-updates-4.2.20-1.fc38.noarch.rpm [Failed to connect to yum.qubes-os.org port 443 after 279 ms: Couldn't connect to server]
[MIRROR] qubes-core-agent-nautilus-4.2.20-1.fc38.x86_64.rpm: Curl error (7): Couldn't connect to server for https://yum.qubes-os.org/r4.2/current/vm/fc38/rpm/qubes-core-agent-nautilus-4.2.20-1.fc38.x86_64.rpm [Failed to connect to yum.qubes-os.org port 443 after 4 ms: Couldn't connect to server]
[FAILED] qubes-core-agent-nautilus-4.2.20-1.fc38.x86_64.rpm: No more mirrors to try - All mirrors were already tried without success
(2-3/6): qubes-cor  5% [-                 ]  47 MB/s |  14 kB     00:00 ETA
The downloaded packages were saved in cache until the next successful transaction.
You can remove cached packages by executing 'dnf clean packages'.
Error: Error downloading packages:
  qubes-core-agent-nautilus-4.2.20-1.fc38.x86_64: Cannot download, all mirrors were already tried without success

```

**Comment by marmarek** [comment]
Can you show earlier messages too? `journalctl -b -u qubes-network-uplink@eth0.service`
Do you have ipv6 enabled? If so, you may want to try update from current-testing (4.2.24) - it fixes related issue.

**Comment by kennethrrosen** [comment]
If ipv6 is enabled by default, then yes. `ip a | grep inet6` shows this to be true. I don't have the technical knowledge exactly to know what has happened, but I can now get networking through the qube, though those two services still fail, along with a third (below). I had  `sys-net --> sys-pihole --> sys-vpn` which worked before the 4.2 upgrade. only sys-vpn had iptables set. 

Then I ran into this issue, but used `qubesctl` to salt a sys-firewall, so now my configuration is `sys-net --> sys-firewall --> sys-pihole --> sys-vpn`. The iptables set in 4.1 in `sys-vpn` were inherited to `nft` in `/rw/config/rc.local` but the only change I made was moving `sys-firwall` into the chain. Again, not sure how this fixed my networking, but the services still fail.

```
Oct 10 16:37:59 core-vpn systemd[1]: Starting qubes-network-uplink@eth0.service - Qubes network uplink (eth0) setup...
Oct 10 16:38:00 core-vpn setup-ip[506]: Error: NetworkManager is not running.
Oct 10 16:38:00 core-vpn setup-ip[511]: Error: any valid address is expected rather than "".
Oct 10 16:38:00 core-vpn systemd[1]: qubes-network-uplink@eth0.service: Main process exited, code=exited, status=1/FAILURE
Oct 10 16:38:00 core-vpn systemd[1]: qubes-network-uplink@eth0.service: Failed with result 'exit-code'.
Oct 10 16:38:00 core-vpn systemd[1]: Failed to start qubes-network-uplink@eth0.service - Qubes network uplink (eth0) setup.
```

```
× qubes-misc-post.service - Qubes misc post-boot actions
     Loaded: loaded (/usr/lib/systemd/system/qubes-misc-post.service; enabled; preset: enabled)
    Drop-In: /usr/lib/systemd/system/service.d
             └─10-timeout-abort.conf
     Active: failed (Result: exit-code) since Tue 2023-10-10 16:38:56 EST; 3min 50s ago
    Process: 572 ExecStart=/usr/lib/qubes/init/misc-post.sh (code=exited, status=1/FAILURE)
   Main PID: 572 (code=exited, status=1/FAILURE)
        CPU: 634ms
```

**Comment by marmarek** [comment]
Do you have `network-manager` service enabled in the vpn qube? (you can check with qvm-service tool, or in the "services" tab in qube settings)

**Comment by kennethrrosen** [comment]
Yes, it is enabled.

**Comment by DemiMarie** [comment]
I think the solution is to drop the systemd service and instead use `RUN` in the udev rules.

**Comment by marmarek** [comment]
> I think the solution is to drop the systemd service and instead use `RUN` in the udev rules.

You sure about that? The above looks like ordering issue (`qubes-network-uplink@eth0.service` configured to use network-manager, called before NetworkManager service is started). Moving it totally outside of systemd service will make such ordering issues _worse_, not better...

But as for the solution here, I'm not sure. we could add `After=NetworkManager.service`, but that will make boot slower for most cases (where NetworkManager isn't used), so I don't like this approach. Maybe we can make the setup-ip script resilient to this race (make it work both when it's called before NM start and after)?

**Comment by DemiMarie** [comment]
> > I think the solution is to drop the systemd service and instead use `RUN` in the udev rules.
> 
> You sure about that? The above looks like ordering issue (`qubes-network-uplink@eth0.service` configured to use network-manager, called before NetworkManager service is started). Moving it totally outside of systemd service will make such ordering issues _worse_, not better...
> 
> But as for the solution here, I'm not sure. we could add `After=NetworkManager.service`, but that will make boot slower for most cases (where NetworkManager isn't used), so I don't like this approach. Maybe we can make the setup-ip script resilient to this race (make it work both when it's called before NM start and after)?

The ordering problems can be worked around by using an explicit `systemctl start network-pre.target network.target  || exit "$?"` command in the script.  NetworkManager should have `Before=network.target` in its own unit file (if not, that’s a bug in NetworkManager or its packaging), so this avoids any ordering problems.  If NetworkManager is _not_ used, then this systemctl command will _not_ wait for `NetworkManager.service`, so the performance penalty should be negligible.

**Comment by marmarek** [comment]
> If NetworkManager is _not_ used, then this systemctl command will _not_ wait for `NetworkManager.service`, so the performance penalty should be negligible.

Unfortunately, this part is simply not true. Systemd evaluates conditions only directly when it's going to start the service, not when deciding on start order.

**Comment by DemiMarie** [comment]
> > If NetworkManager is _not_ used, then this systemctl command will _not_ wait for `NetworkManager.service`, so the performance penalty should be negligible.
> 
> Unfortunately, this part is simply not true. Systemd evaluates conditions only directly when it's going to start the service, not when deciding on start order.

systemd will start the service, yes, but it will finish virtually instantly (because it doesn’t need to do anything).

**Comment by marmarek** [comment]
But it will already delay starting `qubes-network-uplink@eth0.service`, and that's what's where you loose some qube start time.

**Comment by DemiMarie** [comment]
> But it will already delay starting `qubes-network-uplink@eth0.service`, and that's what's where you loose some qube start time.

Ah, so you are talking about the _dependencies_ of `NetworkManager.service`.  That said, it is easy to make this conditional in the shell script: wait for `NetworkManager.service` if it is enabled.

**Comment by marmarek** [comment]
Maybe that's the fix here? At the start of `setup-ip` add "if network-manager enabled, wait for it to start" ? It will artificially prolong `setup-ip` run time, but that should be okay I think (as long as that won't be too long to run into timeouts in service start or udev).
I have some idea how to do "if network-manager enabled" part (`qvsc network-manager && systemctl is-enabled NetworkManager`), but what's the best method to do "wait for it to start" part?

**Comment by DemiMarie** [comment]
`systemctl start NetworkManager.service` should work.

**Comment by DemiMarie** [comment]
That said, I really want to move to `RUN+=` and away from the systemd unit file, because I have had the current code fail more than once on my system (without NetworkManager being involved).  I don’t think systemd units are meant to be started and stopped by udev as Qubes is doing today.

**Comment by marmarek** [comment]
By moving it to just udev rule, you loose possibility to order it against other startup services, like setting up firewall (and generally things that order themselves either before network-pre.target, or after network.target).

**Comment by DemiMarie** [comment]
> By moving it to just udev rule, you loose possibility to order it against other startup services, like setting up firewall (and generally things that order themselves either before network-pre.target, or after network.target).

`systemctl start network-pre.target || exit $?` takes care of `After=network-pre.target` and `Requires=network-pre.target`.  The `Before=network.target` is useless because nothing guarantees that `qubes-network-uplink@eth0.service` is pulled in before `network.target` gets started.

**Comment by marmarek** [comment]
But you do realize you are now trying to re-invent service ordering already done by systemd, outside of it? I just mentioned few cases I remembered, I'm sure I missed several more. In other words, I don't like this move.

**Comment by DemiMarie** [comment]
> But you do realize you are now trying to re-invent service ordering already done by systemd, outside of it?

I do!  If you know of a better approach that doesn’t break occasionally (requiring `systemctl restart qubes-network-uplink@eth0.service` to fix), I would love to know it.

> I just mentioned few cases I remembered, I'm sure I missed several more. In other words, I don't like this move.

The proper, systemd-native solution is to have a persistent daemon that manages network interfaces.  In this model, `qubes-network-uplink@eth0.service` either would not exist (because the daemon would pick up “eth0 has appeared” itself) or would be a `Type=notify` unit, not `Type=oneshot`.  Right now, we are abusing udev to trigger a `Type=oneshot` service that does a specific, non-idempotent action when it is started and again when it is stopped.  This leads to all sorts of race conditions.

**Comment by marmarek** [comment]
Ok, I should have looked at the setup-ip script earlier. It's totally not about NetworkManager running or not. It's about typo that is already fixed in qubes-core-agent 4.2.22: https://github.com/QubesOS/qubes-core-agent-linux/commit/05f5b8c8
